### ISSUE TITLE
CASE expression with 12+ conditions returns null

### ISSUE THREAD

**Issue by mikebin** [confluentinc/ksql#5707]
**Describe the bug**
A `CASE` expression with 12 or more conditions returns a `null` result

**To Reproduce**
ksqlDB versions: 0.9.0, 0.10.0

```sql
CREATE STREAM STR_DEPARTMENTS(DEPT_NO VARCHAR, DEPT_NM VARCHAR) WITH(KAFKA_TOPIC='STR_DEPARTMENTS', VALUE_FORMAT='AVRO', PARTITIONS=1, REPLICAS=1);
INSERT INTO STR_DEPARTMENTS (dept_no, dept_nm) VALUES('1', 'IRS');
INSERT INTO STR_DEPARTMENTS (dept_no, dept_nm) VALUES('2', 'PD');
INSERT INTO STR_DEPARTMENTS (dept_no, dept_nm) VALUES('3', 'JUDICIAL');
INSERT INTO STR_DEPARTMENTS (dept_no, dept_nm) VALUES('4', 'REVENUE');
INSERT INTO STR_DEPARTMENTS (dept_no, dept_nm) VALUES('5', 'MUNCIPAL');
INSERT INTO STR_DEPARTMENTS (dept_no, dept_nm) VALUES('6', 'TELECOM');
INSERT INTO STR_DEPARTMENTS (dept_no, dept_nm) VALUES('7', 'CUSTOMS');
INSERT INTO STR_DEPARTMENTS (dept_no, dept_nm) VALUES('8', 'EPB');
INSERT INTO STR_DEPARTMENTS (dept_no, dept_nm) VALUES('9', 'WATER');
INSERT INTO STR_DEPARTMENTS (dept_no, dept_nm) VALUES('10', 'ENERGY');
INSERT INTO STR_DEPARTMENTS (dept_no, dept_nm) VALUES('11', 'PETROLIUM');
INSERT INTO STR_DEPARTMENTS (dept_no, dept_nm) VALUES('12', 'LAW');
INSERT INTO STR_DEPARTMENTS (dept_no, dept_nm) VALUES('13', 'COMUPTERS');

SELECT CASE WHEN DEPT_NM = 'IRS' THEN '1'
WHEN DEPT_NM = 'PD' THEN '2'
WHEN DEPT_NM = 'JUDICIAL' THEN '3'
WHEN DEPT_NM = 'REVENUE' THEN '4'
WHEN DEPT_NM = 'MUNCIPAL' THEN '5'
WHEN DEPT_NM = 'TELECOM' THEN '6'
WHEN DEPT_NM = 'CUSTOMS' THEN '7'
WHEN DEPT_NM = 'EPB' THEN '8'
WHEN DEPT_NM = 'WATER' THEN '9'
WHEN DEPT_NM = 'ENERGY' THEN '10'
WHEN DEPT_NM = 'PETROLIUM' THEN '11'
WHEN DEPT_NM = 'LAW' THEN '12'
WHEN DEPT_NM = 'COMUPTERS' THEN '13'
END AS TEST
FROM STR_DEPARTMENTS
EMIT CHANGES
;
```
**Expected behavior**
```
+--------------------+
|TEST                |
+--------------------+
|1                   |
|2                   |
|3                   |
|4                   |
|5                   |
|6                   |
|7                   |
|8                   |
|9                   |
|10                  |
|11                  |
|12                  |
|13                  |
```

**Actual behaviour**
```
+--------------------+
|TEST                |
+--------------------+
|null                |
|null                |
|null                |
|null                |
|null                |
|null                |
|null                |
|null                |
|null                |
|null                |
|null                |
|null                |
|null                |
```

Errors appear in the ksqlDB processing log:

```
ksqldb-server       | [2020-06-26 18:30:53,516] ERROR {"type":1,"deserializationError":null,"recordProcessingError":{"errorMessage":"Error computing expression (CASE WHEN (DEPT_NM = 'IRS') THEN '1' WHEN (DEPT_NM = 'PD') THEN '2' WHEN (DEPT_NM = 'JUDICIAL') THEN '3' WHEN (DEPT_NM = 'REVENUE') THEN '4' WHEN (DEPT_NM = 'MUNCIPAL') THEN '5' WHEN (DEPT_NM = 'TELECOM') THEN '6' WHEN (DEPT_NM = 'CUSTOMS') THEN '7' WHEN (DEPT_NM = 'EPB') THEN '8' WHEN (DEPT_NM = 'WATER') THEN '9' WHEN (DEPT_NM = 'ENERGY') THEN '10' WHEN (DEPT_NM = 'PETROLIUM') THEN '11' WHEN (DEPT_NM = 'LAW') THEN '12' WHEN (DEPT_NM = 'COMUPTERS') THEN '13' END) for column TEST with index 0","record":null,"cause":[]},"productionError":null} (processing.6017150808097665617.Project:44)
```

**Additional context**
Works correctly in all cases in ksqlDB 5.5.0 and 5.5.1.

With fewer than 12 conditions, proper results are produced in 0.9.0 and 0.10.0:

```sql
ksql> SELECT CASE WHEN DEPT_NM = 'IRS' THEN '1'
>WHEN DEPT_NM = 'PD' THEN '2'
>WHEN DEPT_NM = 'JUDICIAL' THEN '3'
>WHEN DEPT_NM = 'REVENUE' THEN '4'
>WHEN DEPT_NM = 'MUNCIPAL' THEN '5'
>WHEN DEPT_NM = 'TELECOM' THEN '6'
>WHEN DEPT_NM = 'CUSTOMS' THEN '7'
>WHEN DEPT_NM = 'EPB' THEN '8'
>WHEN DEPT_NM = 'WATER' THEN '9'
>WHEN DEPT_NM = 'ENERGY' THEN '10'
>WHEN DEPT_NM = 'PETROLIUM' THEN '11'
>--WHEN DEPT_NM = 'LAW' THEN '12'
>--WHEN DEPT_NM = 'COMUPTERS' THEN '13'
>END AS TEST
>FROM STR_DEPARTMENTS
>EMIT CHANGES
>;
+--------------------+
|TEST                |
+--------------------+
|1                   |
|2                   |
|3                   |
|4                   |
|5                   |
|6                   |
|7                   |
|8                   |
|9                   |
|10                  |
|11                  |
|null                |
|null                |
```

**Comment by mikebin** [comment]
@big-andy-coates submitted another issue recently with a similar processing error with the `CASE` expression, but it looked like a slightly different scenario. Referencing here in case these issues are related: https://github.com/confluentinc/ksql/issues/5700

**Comment by spena** [comment]
Seems different. This issue does not occur in the master branch (v0.11), but the issue in #5700 occurs in master.

**Comment by big-andy-coates** [comment]
I had a quick look at this...

tl;dr; this is only recreatable when using the docker images.  It likely needs someone with in-depth docker knowledge to progress.

To try to recreate I adding a QTT test for this (See below). This test passed on `master` and earlier community release branches that I could get to build.

Likewise, running the server and cli in the debugger for `master`, and other earlier community releases that I could get to build works, i.e. returns non-null result.

It is only the docker images that return `null`.  Returns null for `0.11.0`, `0.10.1`, `0.10.0`. Didn't try any others.

Very strange!

```json
{
      "name": "large searched case expression",
      "statements": [
        "CREATE STREAM STR_DEPARTMENTS(DEPT_NO VARCHAR, DEPT_NM VARCHAR) WITH(KAFKA_TOPIC='STR_DEPARTMENTS', VALUE_FORMAT='AVRO');",
        "CREATE STREAM OUTPUT AS SELECT CASE WHEN DEPT_NM = 'IRS' THEN '1' WHEN DEPT_NM = 'PD' THEN '2' WHEN DEPT_NM = 'JUDICIAL' THEN '3' WHEN DEPT_NM = 'REVENUE' THEN '4' WHEN DEPT_NM = 'MUNCIPAL' THEN '5' WHEN DEPT_NM = 'TELECOM' THEN '6' WHEN DEPT_NM = 'CUSTOMS' THEN '7' WHEN DEPT_NM = 'EPB' THEN '8' WHEN DEPT_NM = 'WATER' THEN '9' WHEN DEPT_NM = 'ENERGY' THEN '10' WHEN DEPT_NM = 'PETROLIUM' THEN '11' WHEN DEPT_NM = 'LAW' THEN '12' WHEN DEPT_NM = 'COMUPTERS' THEN '13' END AS TEST FROM STR_DEPARTMENTS;"
      ],
      "inputs": [
        {"topic": "STR_DEPARTMENTS", "value": {"DEPT_NO": "1", "DEPT_NM": "IRS"}}
      ],
      "outputs": [
        {"topic": "OUTPUT",  "value": {"TEST": "1"}}
      ]
    }
```

**Comment by big-andy-coates** [comment]
The error in the processing log doesn't have a cause:

Which is less than useful. This is because we remove the first line of the cause here https://github.com/confluentinc/ksql/blob/0.11.x-ksqldb/ksqldb-execution/src/main/java/io/confluent/ksql/logging/processing/RecordProcessingError.java#L143.  For many cases, e.g. NPE, divide by zero, etc, there is only 1 line in the cause.   We should remove this line, better to sometimes have too much _cause_ than not enough.

**Comment by big-andy-coates** [comment]
The error logged to the processing logger doesn't have any `cause` information:

> ksqldb-server       | [2020-06-26 18:30:53,516] ERROR {"type":1,"deserializationError":null,"recordProcessingError":{"errorMessage":"Error computing expression (CASE WHEN (DEPT_NM = 'IRS') THEN '1' WHEN (DEPT_NM = 'PD') THEN '2' WHEN (DEPT_NM = 'JUDICIAL') THEN '3' WHEN (DEPT_NM = 'REVENUE') THEN '4' WHEN (DEPT_NM = 'MUNCIPAL') THEN '5' WHEN (DEPT_NM = 'TELECOM') THEN '6' WHEN (DEPT_NM = 'CUSTOMS') THEN '7' WHEN (DEPT_NM = 'EPB') THEN '8' WHEN (DEPT_NM = 'WATER') THEN '9' WHEN (DEPT_NM = 'ENERGY') THEN '10' WHEN (DEPT_NM = 'PETROLIUM') THEN '11' WHEN (DEPT_NM = 'LAW') THEN '12' WHEN (DEPT_NM = 'COMUPTERS') THEN '13' END) for column TEST with index 0","record":null,"cause":[]},"productionError":null} (processing.6017150808097665617.Project:44)

This is cause the code in the processing logger code explicitly removes the first 'cause' : https://github.com/confluentinc/ksql/blob/0.11.x-ksqldb/ksqldb-execution/src/main/java/io/confluent/ksql/logging/processing/RecordProcessingError.java#L143

This is dropping useful information, and likely hiding the cause of this error.

**Comment by big-andy-coates** [comment]
Update: If I build images locally of the `6.0.0` branch... this all works...

**Comment by stevenpyzhang** [comment]
I'm testing with 0.12.0 images, this is still an issue, the processing log error is now
```
ERROR {"type":1,"deserializationError":null,"recordProcessingError":{"errorMessage":"Error computing expression (CASE WHEN (DEPT_NM = 'IRS') THEN '1' WHEN (DEPT_NM = 'PD') THEN '2' WHEN (DEPT_NM = 'JUDICIAL') THEN '3' WHEN (DEPT_NM = 'REVENUE') THEN '4' WHEN (DEPT_NM = 'MUNCIPAL') THEN '5' WHEN (DEPT_NM = 'TELECOM') THEN '6' WHEN (DEPT_NM = 'CUSTOMS') THEN '7' WHEN (DEPT_NM = 'EPB') THEN '8' WHEN (DEPT_NM = 'WATER') THEN '9' WHEN (DEPT_NM = 'ENERGY') THEN '10' WHEN (DEPT_NM = 'PETROLIUM') THEN '11' WHEN (DEPT_NM = 'LAW') THEN '12' WHEN (DEPT_NM = 'COMUPTERS') THEN '13' END) for column TEST with index 0","record":null,"cause":["Method 'java.util.List java.util.List.of(java.lang.Object[])' must be InterfaceMethodref constant"]},"productionError":null,"serializationError":null} (processing.3436967807172503237.Project:44)
```

**Comment by stevenpyzhang** [comment]
That's really weird @big-andy-coates that if you build the 6.0.0 images locally they don't have this bug, I tried a set up using `confluentinc/cp-ksqldb-server:6.0.0` and the issue is still present

**Comment by cprasad1** [comment]
the problem seems similar to https://stackoverflow.com/questions/52541023/java-lang-incompatibleclasschangeerror-org-junit-jupiter-params-provider-argumen.

They suggest updating gradle/groovy

**Comment by cprasad1** [comment]
someone else ran into a similar issue https://github.com/palantir/gradle-docker/issues/96

**Comment by cprasad1** [comment]
https://github.com/janino-compiler/janino/issues/69

**Comment by cprasad1** [comment]
The version 3.0.7 of janino that we use still produces java6 code and we can't call static interface methods on this code (we have `List.of` somewhere during code generation, idk exactly where/how). They have a fix for this but have't created a release which includes their fix, so we either need to:
1) Build our own binary from their source which includes the latest code and make some (known) tweaks to our code, or
2) Stop generating references to List.of which is exposing the issue. (hard to track down)

**Comment by apurvam** [comment]
Thanks @cprasad1 for the updates. Very interesting findings! Are you saying the java code block we generate for `CASE` statements includes `List.of` statements, which janino 3.0.7 can't compile? And this is hard to track down? If so, why? How are these `List.of` statements generated?

Also, do we know why these CASE statements work with JDK version 8 but not JDK version 11?

**Comment by cprasad1** [comment]
@apurvam there are some new findings. **Exact** cause is still unknown
_Are you saying the java code block we generate for CASE statements includes List.of statements, which janino 3.0.7 can't compile?_
The code block that we generate for CASE statements includes `ImmutableList.of` statements. `ImmutableList.of` has 13 different overloaded definitions. Each one of them has a different number of params (one to 13). The 13th definition that we hit with 13 cases is the variadic one and its behavior is different from the others. Specifically, it makes calls to `new Object[]` and `System.arraycopy()`. Compilation of this code goes fine, but at runtime JDK 11 probably calls some static interface method on which Janino chokes. This is my best guess.

_And this is hard to track down? If so, why?_ 
So far yes. Stepping through Janino code is confusing, and the debugger stops working at one particular point.

_How are these List.of statements generated?_
I still don't know how these `List.of` statements are generated. They are nowhere to be seen except for the error message in the end. JDK 11 probably chooses to execute this method at runtime. Why does it choose to execute this method? I don't know

_Also, do we know why these CASE statements work with JDK version 8 but not JDK version 11?_
Explanations above are just clues for an actual theory. Exact reason is still a mystery. There seems to be a solution tho.
